### PR TITLE
[ES|QL] pow function always returns double

### LINKED ISSUE #99055: [ES|QL] Pow function should always return a double
### Description

Based on the discussion on 8/30, we agreed that Pow should return a double in all cases like the underlying Java Math function does.

### PR DESCRIPTION
**Pull request by not-napoleon**
Resolves https://github.com/elastic/elasticsearch/issues/99055

Relates to https://github.com/elastic/elasticsearch/issues/100558

This corrects an earlier mistake in the ES|QL language design.  Initially we had thought to have `pow` return the same type as its inputs, but in practice even for integer inputs this quickly grows out of the representable range, and we returned null much of the time. This also created a lot of edge cases around casting to/from doubles (which the underlying java function uses).  The version in this PR follows the java spec, by always casting its inputs to doubles, and returning a double. Doing it this way also allows for a rather significant reduction in lines of code.

I removed many of the tests covering pow specific edge cases.  This seems reasonable to me as I expect `java.lang.math.pow` to be well behaved and most of those edge cases were around type testing which no longer applies. At the same time, this simplification allows us to leverage the new scalar function testing framework, which means better null coverage, better type coverage, and much easier extensibility. 

We do consider this a breaking change, but as the feature is still in tech preview and this is a relatively small surface area, we are not too concerned with disruptions.

### CODE REVIEW

**Review by nik9000 — APPROVED** [review]

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/math.csv-spec [hunk 1]
```diff
@@ -335,7 +335,7 @@ ROW base = -4, exponent = 0.5
 // end::powNeg-sqrt[]
 ;
 warning:Line 2:12: evaluation of [POW(base, exponent)] failed, treating result as null. Only first 20 failures recorded.
-warning:Line 2:12: java.lang.ArithmeticException: invalid result: pow(-4.0, 0.5)
+warning:Line 2:12: java.lang.ArithmeticException: invalid result when computing pow
```
Comment (craigtaverner) [inline x-pack/plugin/esql/qa/testFixtures/src/main/resources/math.csv-spec:338]: I presume it is considered a security concern to include parts of the query in the error message? Pity, as this is so much nicer for debugging.

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/math.csv-spec [hunk 2]
```diff
@@ -432,14 +432,10 @@ powULIntOverrun
...
 ;
-// tag::powULOverrun-warning[]
-warning:Line 1:9: evaluation of [POW(9223372036854775808, 2)] failed, treating result as null. Only first 20 failures recorded.
-warning:Line 1:9: java.lang.ArithmeticException: long overflow
-// end::powULOverrun-warning[]
 
 // tag::powULOverrun-result[]
-x:long
-null
+x:double
```
Comment (craigtaverner) [inline x-pack/plugin/esql/qa/testFixtures/src/main/resources/math.csv-spec:437]: If this no longer overruns, we cannot use this in the documentation. Perhaps you need to delete the entire docs section regarding arithmetic overruns?

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/math/Pow.java [hunk 3]
```diff
@@ -70,65 +69,19 @@ public Object fold() {
...
-    static int processInt(double base, double exponent) {
-        return exponent == 1 ? validateAsInt(base) : validateAsInt(base, exponent);
-    }
-
     private static double validateAsDouble(double base, double exponent) {
         double result = Math.pow(base, exponent);
-        if (Double.isNaN(result)) {
-            throw new ArithmeticException("invalid result: pow(" + base + ", " + exponent + ")");
+        if (Double.isNaN(result) || Double.isInfinite(result)) {
+            throw new ArithmeticException("invalid result when computing pow");
```
Comment (craigtaverner) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/math/Pow.java:80]: Sad to see useful debugging info die...

**Review by craigtaverner — CHANGES_REQUESTED** [review]
Looks good, but there is an issue with the docs using an example of arithmetic exceptions from the tests that no longer throw that exception. I think the docs either need to use a different example, or that section of the docs simplified.

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/math.csv-spec (hunk 1 above)
Comment (not-napoleon) [reply x-pack/plugin/esql/qa/testFixtures/src/main/resources/math.csv-spec:338]: It's also very hard to test.  The tests all now inject many values randomly into the execution, and we didn't build a mechanism to also inject those into the warning messages.  We could solve that with some wild cards or such though.

I agree it aids in debugging, and I am also sad to lose it, but it really feels very risky.  If you want to go to infosec and get them to sign off on this type of error message, I'm happy to build test support for doing it and put them back, but absent an affirmative consent from infosec I really think it's a bad idea.

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/math.csv-spec [hunk 4]
```diff
@@ -203,7 +203,7 @@ height:double | s:double
 powSalarySquared
 from employees | eval s = pow(to_long(salary) - 75000, 2) + 10000 | keep salary, s | sort salary desc | limit 4;
 
-salary:integer | s:long
+salary:integer | s:double
```
Comment (dnhatn) [inline x-pack/plugin/esql/qa/testFixtures/src/main/resources/math.csv-spec:206]: @not-napoleon, I have merged the mixed-cluster tests, you might need to add #[skip:-8.11.99] to these tests, as the assertions differ on 8.11.

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/math.csv-spec (hunk 1 above)
Comment (bpintea) [reply x-pack/plugin/esql/qa/testFixtures/src/main/resources/math.csv-spec:338]: Just trying to understand the concern, if the user would simply do a `| KEEP base, exponent`, that'd reveal anyways the values, provided they have the appropriate permissions, no? They can only apply operations on fields they're entitled to.
In any case, we are providing explicit failure messages in most of other places (like numerical overflows, conversions etc.) already. I think it might make sense to either "fix" them all, or keep being explicit for now and address these security concerns holistically.

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/math.csv-spec [hunk 5]
```diff
@@ -432,14 +432,10 @@ powULIntOverrun
 ROW x = POW(9223372036854775808, 2)
 // end::powULOverrun[]
 ;
-// tag::powULOverrun-warning[]
```
[the remark this answers is not in the document]
Comment (not-napoleon) [reply x-pack/plugin/esql/qa/testFixtures/src/main/resources/math.csv-spec:435]: I thought I did delete that from the docs? did I miss something?

[the remark this answers is not in the document]
Comment (not-napoleon) [reply x-pack/plugin/esql/qa/testFixtures/src/main/resources/math.csv-spec:435]: Oh, nevermind, I see it.  I did miss something clearly.

**Comment by not-napoleon** [comment]
@elasticmachine update branch

File: docs/reference/esql/functions/pow.asciidoc [hunk 6]
```diff
@@ -16,62 +17,6 @@ include::{esql-specs}/math.csv-spec[tag=powDI]
...
-include::{esql-specs}/math.csv-spec[tag=powII-result]
-|===
-
-NOTE: The actual power function is performed using double precision values for all cases.
-This means that for very large non-floating point values there is a small chance that the
-operation can lead to slightly different answers than expected.
-However, a more likely outcome of very large non-floating point values is numerical overflow.
-
-[discrete]
-==== Arithmetic errors
```
Comment (craigtaverner) [inline docs/reference/esql/functions/pow.asciidoc:46]: Is it no longer possible to generate an arithmetic error? This section could remain, but with a new example. On the other hand perhaps this is general to all numerical functions, and should be in a generic location.

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/math.csv-spec (hunk 2 above)
Comment (craigtaverner) [reply x-pack/plugin/esql/qa/testFixtures/src/main/resources/math.csv-spec:433]: Since you deleted the docs, you could also delete the tags here.

**Review by craigtaverner — APPROVED** [review]
Looks like you fixed the docs by deleting the arithmetic overrun section. While it might be nice to have that somewhere, I agree it does not need to be here.

File: docs/reference/esql/functions/pow.asciidoc (hunk 6 above)
Comment (not-napoleon) [reply docs/reference/esql/functions/pow.asciidoc:46]: Yeah, it just behaves like the rest of the numeric functions right now.  Also, there are just too many ways to get an error, and too many weird edge cases.  I didn't want to copy all the behavior from the java docs.
